content/post/elm-in-rakuten/index.md

```diff
@@ -0,0 +1,1273 @@
+
+* The **Elm static type system**, discussed in detail below, remove a whole category of surprises related to dynamic typing.
+* The Elm Package Manager supports **Enforced Semantic Versioning**.[^semantic-versioning] There are no surprises in PATCH releases because version numbers are enforced by a script that scans libraries detecting, for example, removal or renaming of exposed functions.
+* **Elm-UI**, a library to render views that translate developers' intents into layouts in a clear manner, discussed in details later.
+
+[^semantic-versioning]: The Elm Package Manager allows you to check differences (functions added/removed/modified) in any published Elm package simply by running the command `elm diff` like in `elm diff elm/json 1.0.0 1.1.2`. Evan Czaplicki gives a live example of this feature in the video [Convergent Evolution](https://youtu.be/jl1tGiUiTtI).
+
+Languages with automatic type conversion, like JavaScript, can lead to surprises.[^funny-youtube-videos] What happens if we add a *number* and a *string* together? For example, `1` and `"2"`? Do we get `3`? Do we get `"12"`? Do we get an error? Do we get something else?
+
+Elm, in contrast, is strongly statically typed but types declaration are optional because the Elm compiler can infer 100% of them, including the types of all external libraries.
+
+[^funny-youtube-videos]: There are several funny videos about this topic on YouTube. One of the most popular is probably [What the... JavaScript?](https://youtu.be/2pL28CcEijU) by Kyle Simpsons
+
+![Dynamic Typing vs Static Typing](https://lucamug.github.io/post-Elm-in-Rakuten/images/dynamic-vs-static.png)
+*Dynamic Typing vs Static Typing, illustration by [Kolja Wilcke](https://twitter.com/01k), licensed under [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/).*
+
+If you need to create a quick proof of concept, dynamic typing may be faster and the puzzle does resemble a giraffe even if it contains mistakes. But for robust applications and correct puzzle solutions, static typing is the right way to go.
+
+TypeScript, which adds optional static typing to JavaScript and is probably one of the best things to happen to JavaScript, can partially mitigate the issues with the JavaScript dynamic type system. But being a superset of JavaScript it needs to compromise on elegance and simplicity. For example, type declarations are optional (*any* as escape hatch), inference doesn’t cover all the code, it requires *type guards*[^type-guards], and not all JavaScript libraries have type annotations.
+
+
+
+[^type-guards]: [Type guards](https://www.typescriptlang.org/docs/handbook/advanced-types.html) are TypeScript expressions that perform a runtime check to discriminate between *custom types*. For example:{{< highlight typescript >}}function isFish(pet: Fish | Bird): pet is Fish {
```

Comment:
Broken code fence

Reply:
from what I can see fences with language break within footnotes, I guess luca resorted to the `highlight` shortcode 🤷‍♀️
I reformatted it a bit so it doesn't look so ugly (apparently hugo allows line breaks, just no blank lines)